### ISSUE TITLE
[question] Copy a package from machine to machine without internet

### ISSUE THREAD

**Issue by WeissFabian** [conan-io/conan#10740]
Hello,

I have two systems one which creates the conan packages (conan export-pkg) and one which needs to install them.
The second system is not connected to the internet at all, so I have to copy the package from one system to the other via a usb stick.
So if the offline system wants to install the package, it can do so without needing a internet connection, since the packages are already on the system.

My question is how do i copy a package from one device to the other, for example with a USB Stick.
Where do i find the package from the first machine to copy to my usb stick and how do i add this package to the offline machine so conan knows its there?

Thanks in advance!

**Comment by memsharded** [comment]
Hi @WeissFabian 

At the moment this is not possible, at least with a Conan built-in mechanism. We are providing some foundations so this might become possible in 2.X, but not in 1.X.

If you really need this now, the workaround could be a plain copy of the cache folders. In Conan 1.X the cache is structured in folders, and most of it is self-contained. Go to your ``<userhome>/.conan/data`` folder and you will find there your packages to copy. It is important to know that this method is not supported and could break with any release (but at the moment in 1.X very unlikely because we are not doing those kind of changes anymore and working mostly on 2.0). Also, there will not be a straightforward upgrade of that in 2.0, because the cache layout is changing and will be less structured, so such a copy will not be evident (might be possible, but not evident).

**Comment by samuel-emrys** [comment]
Crossposting from #13694 for tracking as it was closed as a duplicate of this ticket, to capture the use case and some more detailed requirements.

## Bottom line up front

Given that most CI tools propose the usage of "build artifacts" to transfer data between CI stages, it would be very useful to be able to do two things:
1. Extract a package from the cache in its entirety, including its `exports_sources` and conanfile, and place the contents in a single directory or tarball.
2. Export a package into the cache in its entirety from a single directory or tarball.

The workflow that I'm proposing would look something like:

```bash
$ conan create .
$ conan extract zlib/1.2.13 -pr:b default -pr:h default -o zlib-package # extract package contents to directory `zlib-package`
$ conan export-pkg zlib-package zlib/1.2.13 -pr:b default -pr:h default # place the contents of `zlib-package` back in the cache
```
To illustrate the need for this, I'll first illustrate the problem using a simple GitLab CI pipeline, but it's important to note that this is by no means a problem specific to GitLab. It's a common problem for any CI that relies on connecting separated stages through the use of artifacts. This is also a problem with GitHub Actions and Bamboo.

### The problem

This is a naive build stage.

```
pkg/build:linux:
  stage: build
  image: conanio/${COMPILER}${VERSION}-ubuntu18.04
  parallel:
    matrix:
      - COMPILER: clang
        VERSION: [12, 13, 14]
        LIBCXX: [libstdc++11]
        CPPSTD: [20]
        BUILD_TYPE: [Release, Debug]
      - COMPILER: gcc
        VERSION: [10, 11, 12]
        LIBCXX: [libstdc++11]
        CPPSTD: [20]
        BUILD_TYPE: [Release, Debug]
  script:
    - conan conan create . -s compiler=$COMPILER -s compiler.version=$VERSION -s compiler.libcxx=$LIBCXX -s compiler.cppstd=$CPPSTD -s build_type=$BUILD_TYPE
    - conan upload "*" --confirm -r myremote
```
As you can see, this creates builds for clang 12, 13, 14 and gcc 10, 11, 12 for Debug and Release configurations in parallel. Theres more to set up here and I don't normally configure these via the command line but with profiles, but nonetheless it illustrates the intent. I would also normally only want to upload the package built not the entire cache.

**The problem** with this, is that because all of these different configurations build in parallel, the decision to upload definitionally cannot consider whether other builds in the build matrix have been successful or not, and so it's not possible to conditionally upload based on the success of all configurations in the build matrix.

To clarify, **The goal** here is to make sure that ***all*** configurations in a build matrix are successful before uploading ***any*** packages.

### Previous alternatives

In conan 1.x, this could be solved as follows:

```
pkg/build:linux:
  stage: build
  image: conanio/${COMPILER}${VERSION}-ubuntu18.04
  parallel:
    matrix:
      - COMPILER: clang
        VERSION: [12, 13, 14]
        LIBCXX: [libstdc++11]
        CPPSTD: [20]
        BUILD_TYPE: [Release, Debug]
      - COMPILER: gcc
        VERSION: [10, 11, 12]
        LIBCXX: [libstdc++11]
        CPPSTD: [20]
        BUILD_TYPE: [Release, Debug]
  script:
    - conan install . -s compiler=$COMPILER -s compiler.version=$VERSION -s compiler.libcxx=$LIBCXX -s compiler.cppstd=$CPPSTD -s build_type=$BUILD_TYPE
    - conan build .
    - source build/$BUILD_TYPE/generators/conanbuild.sh
    - cmake --install build/$BUILD_TYPE --prefix $(pwd)/package/${COMPILER}-${VERSION}-${LIBCXX}-${BUILD_TYPE}
    - source build/$BUILD_TYPE/generators/deactivate_conanbuild.sh
  artifacts:
    paths:
      - package/

pkg/deploy:linux:
  stage: deploy
  image: conanio/${COMPILER}${VERSION}-ubuntu18.04
  parallel:
    matrix:
      - COMPILER: clang
        VERSION: [12, 13, 14]
        LIBCXX: [libstdc++11]
        CPPSTD: [20]
        BUILD_TYPE: [Release, Debug]
      - COMPILER: gcc
        VERSION: [10, 11, 12]
        LIBCXX: [libstdc++11]
        CPPSTD: [20]
        BUILD_TYPE: [Release, Debug]
  needs:
    - job: pkg/build:linux
      artifacts: true
  script:
    - export PACKAGE_NAME="$(conan inspect . --raw name)"
    - export PACKAGE_VERSION="$(conan inspect . --raw version)"
    - conan export-pkg . $PACKAGE_NAME/$PACKAGE_VERSION@user/stable -pf package/${COMPILER}-${VERSION}-${LIBCXX}-${BUILD_TYPE} -pr:b default -pr:h default
    - conan upload $PACKAGE_NAME/$PACKAGE_VERSION@user/stable -r myremote --confirm
```

What this is essentially doing is manually constructing the package folders for each of the configurations in the build stage and adding them to the `package/` directory such that there would be a directory named, i.e. `package/clang-12-Release`. This `package/` directory is then exported as an artifact from the `build` stage.

The `deploy` stage has a dependency on the `build` stage, and consumes the `package/` artifact. For each configuration, it executes `export-pkg` and then `conan upload`.

This is a reasonably nice solution because there's no time or space inefficiencies - I don't have to rebuild the package after I've confirmed all packages can build, and I don't have to pass unnecessary files around as artifacts. The downside to this solution is that it ***doesn't*** package the `exports_sources`, which means that whilst pre-built binaries for the uploaded configurations are made available on the remote, unless another step is added, the package is unable to be built from sources in the case of a missing configuration.

Of course, this relies on legacy conan 1.x behaviour in which one was able to call `conan package .` and create a package folder in the local directory, and then consume the package folder using `export-pkg`, which is not available in conan 2.0 anymore. Even in modern 1.x, this approach is a bit of a stretch because it's using deprecated methods.

### Current alternatives

So, what are the current alternatives to address this problem?

1. **Accept that `upload` and `create` can't be separated**. Introducing a predicate to ensure that any failures in the pipeline prevent the movement of a package from a feature branch to a stable branch can be achieved by ensuring that the pipeline doesn't progress if there are any failures in the build matrix. Packages from each build will get uploaded to a temporary artifactory repository, which get copied to a stable repository if all builds are successful.
2. **Export the entire conan cache as an artifact from the build stage.** This will likely take a lot of space and pretty space inefficient. I could attempt to trim this down by removing every package from the cache except the package being built, but it's still a pretty messy fix.
3. **Run `conan create` in the build stage, and `conan create && conan upload` in the `deploy` stage.** This builds the package twice for each configuration. It doubles the CI evaluation time, which is reasonably significant in the case of large libraries. 
4. **Build the configurations in the build matrix serially**. This would allow one to only execute `conan upload` once every package has successfully built. The downside to this is that for large numbers of configurations, this could take a substantial amount of time.

Unfortunately option (1) above isn't practical for us in a corporate context, where permissions to create and delete repositories on our corporate artifactory instance are tightly controlled. This is a non starter, and so we're forced to look at the more conventional approaches to this problem where artifactory can't act as a middle man to tie stages together, and more towards the solutions most CI tools provide for this issue.

## The request

So, this brings us to the request. I think that given that this is the paradigm that a large number of CI tools adopt, it would be very useful to be able to do two things:
1. Extract a package from the cache in its entirety, including its `exports_sources` and conanfile, and place the contents in a single directory or tarball.
2. Export a package into the cache in its entirety from a single directory or tarball.

The workflow that I'm proposing would look something like:

```bash
$ conan create .
$ conan extract zlib/1.2.13 -pr:b default -pr:h default -o zlib-package # extract package contents to directory `zlib-package`
$ conan export-pkg zlib-package zlib/1.2.13 -pr:b default -pr:h default # place the contents of `zlib-package` back in the cache
```

Importantly, ***`conan extract` also adds `exports_sources` to the directory*** so that a package can still be compiled for a configuration not contained in the pre-build matrix. This would give a pipeline that could be constructed simply, as follows:

```
pkg/build:linux:
  stage: build
  image: conanio/${COMPILER}${VERSION}-ubuntu18.04
  parallel:
    matrix:
      - COMPILER: clang
        VERSION: [12, 13, 14]
        LIBCXX: [libstdc++11]
        CPPSTD: [20]
        BUILD_TYPE: [Release, Debug]
      - COMPILER: gcc
        VERSION: [10, 11, 12]
        LIBCXX: [libstdc++11]
        CPPSTD: [20]
        BUILD_TYPE: [Release, Debug]
  script:
    - export PACKAGE_NAME="$(conan inspect . --query name)"
    - export PACKAGE_VERSION="$(conan inspect . --query version)"
    - conan create . -pr:b default -pr:h default
    - conan extract $PACKAGE_NAME/$PACKAGE_VERSION -pr:b default -pr:h default -o $(pwd)/package/${COMPILER}-${VERSION}-${LIBCXX}-${BUILD_TYPE}
  artifacts:
    paths:
      - package/

pkg/deploy:linux:
  stage: deploy
  image: conanio/${COMPILER}${VERSION}-ubuntu18.04
  parallel:
    matrix:
      - COMPILER: clang
        VERSION: [12, 13, 14]
        LIBCXX: [libstdc++11]
        CPPSTD: [20]
        BUILD_TYPE: [Release, Debug]
      - COMPILER: gcc
        VERSION: [10, 11, 12]
        LIBCXX: [libstdc++11]
        CPPSTD: [20]
        BUILD_TYPE: [Release, Debug]
  needs:
    - job: pkg/build:linux
      artifacts: true
  script:
    - export PACKAGE_NAME="$(conan inspect . --query name)"
    - export PACKAGE_VERSION="$(conan inspect . --query version)"
    - conan export-pkg package/${COMPILER}-${VERSION}-${LIBCXX}-${BUILD_TYPE} $PACKAGE_NAME/$PACKAGE_VERSION@user/stable -pr:b default -pr:h default
    - conan upload $PACKAGE_NAME/$PACKAGE_VERSION@user/stable -r myremote --confirm
```

If there are other suggestions anybody has about ways in which the problem can be tackled that I haven't considered, very open to considering them.

If you have any questions about this please ask them - this has been a lengthy post and there's a lot here to digest. I'm sure I won't have communicated everything as clearly as I hope to have.

**Comment by zillemarco** [comment]
I totally agree with @samuel-emrys request above.

We have the same situation in which we have a build matrix in our CI pipeline, which should be uploaded to the registry only once all build jobs have passed and we should not have to "rebuild" the code in two different stages (one of the workarounds mentioned by Samuel).

@memsharded I am wondering, how did you approach this issue for Conan center? I suppose you have a similar setup? 🤔

**Comment by zillemarco** [comment]
This is just a brain dump, but would it this be achievable by adding an option to `conan create` to tell it where to create the package (instead of going to the local cache) and another option to `conan upload` to tell is where to find the package to upload? 🤔

**Comment by samuel-emrys** [comment]
> @memsharded I am wondering, how did you approach this issue for Conan center? I suppose you have a similar setup? thinking

I'm reasonably certain that conan center uses the first solution:

> 1. **Accept that upload and create can't be separated.** Introducing a predicate to ensure that any failures in the pipeline prevent the movement of a package from a feature branch to a stable branch can be achieved by ensuring that the pipeline doesn't progress if there are any failures in the build matrix. Packages from each build will get uploaded to a temporary artifactory repository, which get copied to a stable repository if all builds are successful.

**Comment by memsharded** [comment]
Hi @zillemarco 

There are 2 different things here.
This feature is planned, and actually most of the work preparing the packages zipping was already done in the Conan 2.0 "upload" internals.

A different thing is:
> We have the same situation in which we have a build matrix in our CI pipeline, which should be uploaded to the registry only once all build jobs have passed and we should not have to "rebuild" the code in two different stages (one of the workarounds mentioned by Samuel).

The recommended approach for this is not storing the packages in disk, but putting them in a temporary or "build" server repo. Not the main "development" repo, but a separate one, so it doesn't interfere. This approach has several advantages, like simplifying the reproduction from developers machine, traceability, and more easily distributing builds. This is the approach we implement in ConanCenter to build users PRs. In our case it is also important, to avoid re-builds when something breaks, as building too many configuration you cannot just rebuild 129 binaries, just because the machine building the 130th configuration died. Having the packages in actual repos that we can ``conan list``, ``conan install``, ``conan test`` separately, it is critical.


So the feature is planned for the other exceptional scenarios, like having to move some packages over a fully air-gapped environment, when it is not possible to just pull them from a common server, even temporarily (for scenarios where it is possible to populate the internal repo by momentarily allowing access to a controlled server Conan repo, that would still be recommended). It is also possible that some servers like Artifactory would already implement their own "backup and restore" functionality that allows to do this directly on the server, which will probably be much faster than doing it via the client.

**Comment by zillemarco** [comment]
Thanks for the feedback @memsharded @samuel-emrys 🙇 

Just to make sure I understood correctly, what you mean is to have a "staging" server where the CI always uploads to then, when all jobs passed, the actual "upload" job would install the packages from the staging environment and push them to the production one?

Like having two remotes and migrate packages from one remote to the other, using the CI pipeline as "transfering" tool? 🤔

**Comment by memsharded** [comment]
> Just to make sure I understood correctly, what you mean is to have a "staging" server where the CI always uploads to then, when all jobs passed, the actual "upload" job would install the packages from the staging environment and push them to the production one?

Yes, exactly. Even better, for Artifactory server, it is not necessary to download and then upload to do the "promotion" work. There are APIs and this can be done in the server directly from repo to repo, that happens very quickly, because due to file-deduplication, there is actually no copy done or need for extra storage, plus being "more atomic".

**Comment by zillemarco** [comment]
Just wanted to share my experiments 🙂 Yesterday I tried to "trick the system" but it didn't work 😅 

What I tried to do is this:
- On each CI job that builds the package
    - Create a file named `global.conf` to set the value of `core.cache:storage_path` to a local cache folder easy enough to access (this will contain the artifacts to pass to later jobs)
    - Use `conan config install` to install the `gobal.conf` file just created
    - Use `conan create ...` to create the package
    - Copy to the artifacts only the folder of the package build (for examlpe, if the package is for `mylib` I would have copied `./conan_ci/cache/mylib*/**`)
- On the upload job, once all build jobs are finished
    - Set up the local cache just as above (customizing the `global.conf` file)
    - Login to the remote registry
    - Use `conan upload ...` to upload the packages inside the cache to the registry

But this didn't work, mainly because I am not passing the `cache.sqlite3` database from the build jobs to the upload job, and I can't even do that because it wouldn't merge the different databases into one, just override it.

One workaround could be to have one upload job per build job, that way I can add the `cache.sqlite3` to the artifacts, but this is really a hack. Pushing to one registry during build and "moving" to the production registry upon success is preferred, until we can actually have official support for this workflow 🙂

**Comment by memsharded** [comment]
Implemented in https://github.com/conan-io/conan/pull/14923, for next 2.0.14 release.
